renderdoc/driver/metal/metal_types.cpp

```diff
@@ -0,0 +1,50 @@
+/******************************************************************************
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2022 Baldur Karlsson
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#include "metal_types.h"
+
+RDCCOMPILE_ASSERT(sizeof(NS::Integer) == sizeof(std::intptr_t), "NS::Integer size does not match");
+RDCCOMPILE_ASSERT(sizeof(NS::UInteger) == sizeof(std::uintptr_t),
+                  "NS::UInteger size does not match");
+
+template <typename SerialiserType>
+void DoSerialise(SerialiserType &ser, NS::String *&el)
+{
+  const char *cStr = el ? el->utf8String() : NULL;
```

Comment:
You should construct an `rdcstr` here. If you construct a C string the serialiser will have to cache it for its entire lifetime (since it doesn't know what the lifetime of the string is, it ensures it is valid indefinitely). If you serialise via an `rdcstr` it will just be a temporary for the sake of serialisation.

Reply:
Thank you for a good spot. As the c-string is being copied into an NS::String could/should have been freeing the data anyway.

Something like this?
```
template <typename SerialiserType>
void DoSerialise(SerialiserType &ser, NS::String *&el)
{
  rdcstr rdcStr;
  if(el)
  {
    rdcStr = el->utf8String();
  }
  SERIALISE_ELEMENT(rdcStr);

  if(ser.IsReading())
  {
    el = NS::String::string(rdcStr.data(), NS::UTF8StringEncoding);
  }
}
```

After the change

<img width="494" alt="image" src="https://user-images.githubusercontent.com/39392/158048507-0ecb93bb-f578-4b3a-a8db-26950feb95f5.png">

Reply:
Yep that's exactly what I had in mind.

Reply:
d62879c53e9e464f4ba383b323c9783e2e056eec

Reply:
9d0656f5d31a88dbff9bd5f3cd344699e45f8a56